File: apps/backend/src/modules/lawyer/infrastructure/persistence/lawyer.repository.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, In } from 'typeorm';
import {
  ILawyerRepository,
  LawyerSearchCriteria,
} from '../../domain/repositories/lawyer.repository.interface';
import { Lawyer } from '../../domain/entities';
import { SpecializationType, LawyerStatus } from '../../domain/enums';
import { LawyerOrmEntity } from './lawyer.orm-entity';
import { LawyerMapper } from './lawyer.mapper';
import { v4 as uuidv4 } from 'uuid';

/**
 * LawyerRepository
 *
 * TypeORM implementation of ILawyerRepository
 */
@Injectable()
export class LawyerRepository implements ILawyerRepository {
  constructor(
    @InjectRepository(LawyerOrmEntity)
    private readonly repository: Repository<LawyerOrmEntity>,
  ) {}

  async nextId(): Promise<string> {
    return uuidv4();
  }

  async findById(id: string): Promise<Lawyer | null> {
    const ormEntity = await this.repository.findOne({ where: { id } });
    if (!ormEntity) return null;
    return LawyerMapper.toDomain(ormEntity);
  }

  async findByUserId(userId: string): Promise<Lawyer | null> {
    const ormEntity = await this.repository.findOne({ where: { userId } });
    if (!ormEntity) return null;
    return LawyerMapper.toDomain(ormEntity);
  }

  async findByLicenseNumber(licenseNumber: string): Promise<Lawyer | null> {
    const ormEntity = await this.repository.findOne({
      where: { licenseNumber },
    });
    if (!ormEntity) return null;
    return LawyerMapper.toDomain(ormEntity);
  }

  async search(criteria: LawyerSearchCriteria): Promise<{
    lawyers: Lawyer[];
    total: number;
  }> {
    const query = this.repository.createQueryBuilder('lawyer');

    // Filter by status
    if (criteria.status) {
      query.andWhere('lawyer.status = :status', { status: criteria.status });
    }

    // Filter by availability
    if (criteria.isAvailable !== undefined) {
      query.andWhere('lawyer.isAvailable = :isAvailable', {
        isAvailable: criteria.isAvailable,
      });
    }

    // Filter by minimum rating
    if (criteria.minRating) {
      query.andWhere('lawyer.ratingValue >= :minRating', {
        minRating: criteria.minRating,
      });
    }

    // Filter by minimum experience
    if (criteria.minExperience) {
      query.andWhere('lawyer.experienceYears >= :minExperience', {
        minExperience: criteria.minExperience,
      });
    }

    // Filter by specializations (at least one match)
    if (criteria.specializations && criteria.specializations.length > 0) {
      const specializationConditions = criteria.specializations
        .map((spec) => `lawyer.specializations LIKE '%${spec}%'`)
        .join(' OR ');
      query.andWhere(`(${specializationConditions})`);
    }

    // Order by rating and experience
    query.orderBy('lawyer.ratingValue', 'DESC');
    query.addOrderBy('lawyer.experienceYears', 'DESC');

    // Pagination
    if (criteria.limit) {
      query.limit(criteria.limit);
    }
    if (criteria.offset) {
      query.offset(criteria.offset);
    }

    const [ormEntities, total] = await query.getManyAndCount();
    const lawyers = ormEntities.map((entity) => LawyerMapper.toDomain(entity));

    return { lawyers, total };
  }

  async findPendingVerification(limit?: number): Promise<Lawyer[]> {
    const query = this.repository
      .createQueryBuilder('lawyer')
      .where('lawyer.verificationStatus IN (:...statuses)', {
        statuses: ['pending', 'in_review', 'documents_requested'],
      })
      .orderBy('lawyer.createdAt', 'ASC');

    if (limit) {
      query.limit(limit);
    }

    const ormEntities = await query.getMany();
    return ormEntities.map((entity) => LawyerMapper.toDomain(entity));
  }

  async save(lawyer: Lawyer): Promise<void> {
    const ormEntity = LawyerMapper.toOrm(lawyer);
    await this.repository.save(ormEntity);
  }

  async delete(id: string): Promise<void> {
    await this.repository.delete(id);
  }

  async existsByLicenseNumber(licenseNumber: string): Promise<boolean> {
    const count = await this.repository.count({
      where: { licenseNumber },
    });
    return count > 0;
  }

  async findBySpecialization(
    specialization: SpecializationType,
    limit?: number,
  ): Promise<Lawyer[]> {
    const query = this.repository
      .createQueryBuilder('lawyer')
      .where('lawyer.specializations LIKE :spec', {
        spec: `%${specialization}%`,
      })
      .andWhere('lawyer.status = :status', { status: LawyerStatus.Active })
      .andWhere('lawyer.isAvailable = :available', { available: true })
      .orderBy('lawyer.ratingValue', 'DESC');

    if (limit) {
      query.limit(limit);
    }

    const ormEntities = await query.getMany();
    return ormEntities.map((entity) => LawyerMapper.toDomain(entity));
  }

  async findTopRated(limit: number = 10): Promise<Lawyer[]> {
    const ormEntities = await this.repository.find({
      where: {
        status: LawyerStatus.Active,
        isAvailable: true,
      },
      order: {
        ratingValue: 'DESC',
        reviewCount: 'DESC',
      },
      take: limit,
    });

    return ormEntities.map((entity) => LawyerMapper.toDomain(entity));
  }
}

File: apps/backend-python/app/modules/lawyer/domain/repositories/lawyer_repository.py
"""
ILawyerRepository Interface

Абстрактный интерфейс для работы с хранилищем юристов.
"""

from abc import ABC, abstractmethod
from typing import List, Optional

from ..entities.lawyer import Lawyer
from ..value_objects.specialization import SpecializationType
from ..value_objects.verification_status import VerificationStatusType


class ILawyerRepository(ABC):
    """
    Repository interface для Lawyer aggregate.

    Определяет контракт для работы с хранилищем юристов.
    Реализация находится в Infrastructure Layer.

    Methods:
        save: Сохранить юриста
        find_by_id: Найти по ID
        find_by_user_id: Найти по user_id
        find_by_status: Найти по статусу верификации
        search: Поиск юристов с фильтрами
        exists_by_user_id: Проверить существование по user_id
        delete: Удалить юриста
        get_top_rated: Получить топ юристов по рейтингу
        count_by_status: Подсчитать количество по статусу
    """

    @abstractmethod
    async def save(self, lawyer: Lawyer) -> None:
        """
        Сохраняет юриста (create или update).

        Args:
            lawyer: Lawyer entity для сохранения
        """
        pass

    @abstractmethod
    async def find_by_id(self, lawyer_id: str) -> Optional[Lawyer]:
        """
        Находит юриста по ID.

        Args:
            lawyer_id: ID юриста

        Returns:
            Lawyer entity или None
        """
        pass

    @abstractmethod
    async def find_by_user_id(self, user_id: str) -> Optional[Lawyer]:
        """
        Находит юриста по user_id.

        Args:
            user_id: ID пользователя из Identity Module

        Returns:
            Lawyer entity или None
        """
        pass

    @abstractmethod
    async def find_by_status(
        self,
        status: VerificationStatusType,
        limit: int = 100,
        offset: int = 0,
    ) -> List[Lawyer]:
        """
        Находит юристов по статусу верификации.

        Args:
            status: Статус верификации
            limit: Максимальное количество результатов
            offset: Смещение для пагинации

        Returns:
            Список Lawyer entities
        """
        pass

    @abstractmethod
    async def search(
        self,
        specializations: Optional[List[SpecializationType]] = None,
        min_rating: Optional[float] = None,
        max_price: Optional[float] = None,
        location: Optional[str] = None,
        is_available: Optional[bool] = None,
        min_experience: Optional[int] = None,
        query: Optional[str] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> tuple[List[Lawyer], int]:
        """
        Поиск юристов с фильтрами.

        Args:
            specializations: Фильтр по специализациям
            min_rating: Минимальный рейтинг
            max_price: Максимальная цена
            location: Город/регион
            is_available: Доступность
            min_experience: Минимальный опыт (годы)
            query: Текстовый поиск (по имени, описанию)
            limit: Максимальное количество результатов
            offset: Смещение для пагинации

        Returns:
            Tuple (список Lawyer entities, общее количество)
        """
        pass

    @abstractmethod
    async def exists_by_user_id(self, user_id: str) -> bool:
        """
        Проверяет существование юриста по user_id.

        Args:
            user_id: ID пользователя

        Returns:
            True если существует, False иначе
        """
        pass

    @abstractmethod
    async def delete(self, lawyer_id: str) -> None:
        """
        Удаляет юриста.

        Args:
            lawyer_id: ID юриста для удаления
        """
        pass

    @abstractmethod
    async def get_top_rated(
        self,
        specialization: Optional[SpecializationType] = None,
        location: Optional[str] = None,
        limit: int = 10,
    ) -> List[Lawyer]:
        """
        Получает топ юристов по рейтингу.

        Args:
            specialization: Фильтр по специализации (опционально)
            location: Фильтр по локации (опционально)
            limit: Количество результатов

        Returns:
            Список Lawyer entities (sorted by rating DESC)
        """
        pass

    @abstractmethod
    async def count_by_status(self, status: VerificationStatusType) -> int:
        """
        Подсчитывает количество юристов по статусу.

        Args:
            status: Статус верификации

        Returns:
            Количество юристов
        """
        pass

    @abstractmethod
    async def get_all_pending(self, limit: int = 100) -> List[Lawyer]:
        """
        Получает всех юристов в статусе PENDING (для админ панели).

        Args:
            limit: Максимальное количество

        Returns:
            Список Lawyer entities
        """
        pass
